src/snowflake/cli/_plugins/nativeapp/sf_sql_facade.py

```diff
@@ -0,0 +1,199 @@
+        try:
+            yield
+        finally:
+            if is_different_wh and prev_wh is not None:
+                self._log.debug(f"Switching back to warehouse:{prev_wh}")
+                self._use_object(ObjectType.WAREHOUSE, prev_wh)
+
+    @contextmanager
+    def _use_role_optional(self, new_role: str | None):
+        """
+        Switches to a different role for a while, then switches back.
+        This is a no-op if the requested role is already active or if no role is passed in.
+        @param new_role: Name of the role to use. If not a valid Snowflake identifier, will be converted before use.
+        """
+        if new_role is None:
+            yield
+            return
+
+        valid_role_name = to_identifier(new_role)
+        try:
+            prev_role_res = self._sql_executor.execute_query(
+                "select current_role()"
+            ).fetchone()
+        except Exception as err:
+            return handle_unclassified_error(err, "Failed to select current role.")
+
+        prev_role = prev_role_res[0]
+        is_different_role = valid_role_name.lower() != prev_role.lower()
+        if is_different_role:
+            self._log.debug(f"Assuming different role: {valid_role_name}")
+            self._use_object(ObjectType.ROLE, valid_role_name)
+        try:
+            yield
+        finally:
+            if is_different_role:
+                self._log.debug(f"Switching back to role:{prev_role}")
+                self._use_object(ObjectType.ROLE, prev_role)
+
+    @contextmanager
+    def _use_database_optional(self, database_name: str | None):
+        """
+        Switch to database `database_name`, then switches back.
+        This is a no-op if the requested database is already selected or if no database_name is passed in.
+        @param database_name: Name of the database to use. If not a valid Snowflake identifier, will be converted before use.
+        """
+
+        if database_name is None:
+            yield
+            return
+
+        valid_name = to_identifier(database_name)
+
+        try:
+            db_result = self._sql_executor.execute_query(
+                "select current_database()"
+            ).fetchone()
+        except Exception as err:
+            return handle_unclassified_error(err, "Failed to select current database.")
+
+        try:
+            prev_db = db_result[0]
+        except:
+            prev_db = None
+
+        is_different_db = valid_name != prev_db
+        if is_different_db:
+            self._log.debug(f"Using database {valid_name}")
+            self._use_object(ObjectType.DATABASE, valid_name)
+
+        try:
+            yield
+        finally:
+            if is_different_db and prev_db is not None:
+                self._log.debug(f"Switching back to database:{prev_db}")
+                self._use_object(ObjectType.DATABASE, prev_db)
+
+    def execute_user_script(
+        self,
+        queries: str,
+        script_name: str,
```

Comment:
should it be a Path? What should it be relative to?

Reply:
this func uses the script_name for log/error purposes (not to read from disk, since template expansion etc happens before).
As for its type, the sql script paths are represented by `str` types right now, both in `apply_package_script` that I have used as an example here (though this function is never reached anymore and will be removed, beside the point) and in post_deploy_hooks, see `SqlScriptHookType`.

Reply:
We shouldn't, in general, represent paths as strings. Not a blocker here, but we should convert incoming strings to paths, and standardize how we refer to project artifacts across the board. IMHO this API should eventually accept something like a relative script path (relative to the project directory).